pex/cli/commands/docs.py

```diff
@@ -0,0 +1,220 @@
+                return False
+            raise
+
+    def kill(self):
+        # type: () -> None
+        os.kill(self.pid, signal.SIGTERM)
+
+
+@attr.s(frozen=True)
+class LaunchResult(object):
+    url = attr.ib()  # type: str
+    already_running = attr.ib()  # type: bool
+
+
+def launch_docs_server(
+    document_root,  # type: str
+    port,  # type: int
+    timeout=5.0,  # type: float
+):
+    # type: (...) -> Union[str, LaunchResult]
+
+    pidfile = Pidfile.load()
+    if pidfile and pidfile.alive():
+        return LaunchResult(url=pidfile.url, already_running=True)
+
+    # Not proper daemonization, but good enough.
+    log = os.path.join(SERVER_DIR, "log.txt")
+    http_server_module = "http.server" if sys.version_info[0] == 3 else "SimpleHttpServer"
+    env = os.environ.copy()
+    # N.B.: We set up line buffering for the process pipes as well as the underlying Python running
+    # the http server to ensure we can observe the `Serving HTTP on ...` line we need to grab the
+    # ephemeral port chosen.
+    env.update(PYTHONUNBUFFERED="1")
+    with safe_open(log, "w") as fp:
+        process = subprocess.Popen(
+            args=[sys.executable, "-m", http_server_module, str(port)],
+            env=env,
+            cwd=document_root,
+            preexec_fn=os.setsid,
+            bufsize=1,
+            stdout=fp.fileno(),
+            stderr=subprocess.STDOUT,
+        )
+
+    pidfile = Pidfile.record(server_log=log, pid=process.pid, timeout=timeout)
+    if not pidfile:
+        try:
+            os.kill(process.pid, signal.SIGKILL)
+        except OSError as e:
+            if e.errno != errno.ESRCH:  # No such process.
+                raise
+        return log
+
+    return LaunchResult(url=pidfile.url, already_running=False)
+
+
+def shutdown_docs_server():
+    # type: () -> bool
+
+    pidfile = Pidfile.load()
+    if not pidfile:
+        return False
+
+    logger.info(
+        "Killing {server} {url} @ {pid}".format(
+            server=SERVER_NAME, url=pidfile.url, pid=pidfile.pid
+        )
+    )
+    pidfile.kill()
+    return True
+
+
+class Docs(BuildTimeCommand):
+    """Interact with the Pex documentation."""
+
+    def run(self):
+        # type: () -> Result
+        html_docs = docs.root(doc_type="html")
+        if not html_docs:
+            # TODO(John Sirois): Evaluate if this should fall back to opening latest html docs instead of just
+            #  displaying links.
+            return Error(
+                dedent(
+                    """\
+                    This Pex distribution does not include embedded docs.
+
+                    You can find the latest docs here:
+                    HTML: https://docs.pex-tool.org
+                     PDF: https://github.com/pex-tool/pex/releases/latest/download/pex.pdf
+                    """
+                ).rstrip()
+            )
+
+        # TODO(John Sirois): Consider trying a standard pex docs port, then fall back to ephemeral if:
```

Comment:
As prior art, if it's interesting: some tools I've seen will attempt use a standard port and then try port+1, port+2 etc. if that doesn't work (with a user prompt if they want to use that new port).

Reply:
Ok, knocked off this TODO in #2365.